counter: mcux_lptmr: Fix prescaler/glitch filter configuration

Holt-Sun opened the pull request:
and frequency calculation

- Change prescale-glitch-filter-bypass from boolean to int enum [0, 1] to align with hardware register representation
- Fix prescaler division calculation in time counter mode to use 2^(value + 1) instead of 2^value
- Fix glitch filter calculation in pulse counter mode to use 2^value
- Introduce MCUX_LPTMR_EFFECTIVE_FREQ macro to correctly calculate counter frequency based on mode and bypass settings
- Remove incorrect prescaler_glitch offset calculation that added timer_mode_sel - 1
- Update prescale-glitch-filter enum range from [0-16] to [0-15] as pulse mode cannot support value 16
- Improve BUILD_ASSERT error message for prescale-glitch-filter range
- Update devicetree binding documentation to clarify prescaler and glitch filter behavior

This fixes incorrect counter frequency reporting and ensures the prescaler/glitch filter values are correctly mapped to hardware register settings according to the NXP LPTMR reference manual.

Holt-Sun committed:
counter: mcux_lptmr: fix prescaler/glitch filter configuration

Fix incorrect LPTMR prescaler/glitch filter mapping that led to wrong
frequency calculation and wrong hardware configuration.

- Calculate effective counter frequency correctly:
  * Time Counter mode: divide by 2^(value + 1)
  * Pulse Counter mode: divide by 2^value
- Map prescale-glitch-filter directly to the HAL enum (no offset math)
- Add prescale-glitch-filter-bypass DT boolean (default false)
- Restrict prescale-glitch-filter to 0..15 and update bindings/DTS users
- Add build-time validation for Pulse mode (value 0 requires bypass)

Signed-off-by: Holt Sun <holt.sun@nxp.com>

Holt-Sun commented:
Add @zejiang0jason as assignee because he is the collaborator of NXP platform drivers and take lots of routine maintenance work.

zejiang0jason commented on dts/bindings/counter/nxp,lptmr.yaml:
```diff
@@ -52,21 +52,27 @@ properties:
+      When in pulse mode, if prescale-glitch-filter is 0, the glitch filter does not
+        support it. If prescale-glitch-filter is greater than 0, glitch filter
+        recognizes change on input pin after 2^prescale-glitch-filter rising clock
+        edges
```
is it better to mention this change in release notes?

Holt-Sun committed:
doc: migration-guide-4.4: add NXP LPTMR counter updates

Document breaking changes in prescale-glitch-filter configuration
and the new prescale-glitch-filter-bypass property. Provide migration
example for out-of-tree device trees.

Signed-off-by: Holt Sun <holt.sun@nxp.com>

lucien-nxp commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -93,6 +93,46 @@ Controller Area Network (CAN)
+     /* New (correct) */
+     lptmr0: counter@40040000 {
+         compatible = "nxp,lpc-lptmr";
+         prescale-glitch-filter-bypass; /* Explicitly bypassed */
```
I have a question here: if user define the prescale-glitch-filter-bypass and  prescale-glitch-filter two parameter at the same time, should we generate a error log for users about the incorrect configuration? Because I bypassed this prescale-glitch-filter at the SOC level.

lucien-nxp replied:
Then could we provide a few examples to show how to use prescale-glitch-filter?  It seems that we have bypassed the   prescaler/glitch filte default on all the platforms. These driver upgrades are actually incompatible with the previous user's usage habits. After upgrading the Zephyr version, they may actually have an impact on users.

Holt-Sun replied:
Here are the updates:
1. Kept the BUILD_ASSERT in counter_mcux_lptmr.c so any out-of-tree DTS that uses Pulse mode with prescale-glitch-filter = <0> and no prescale-glitch-filter-bypass; will fail the build and must be migrated.
2. Ensured bypass default remains false when the boolean property is absent (so prescale-glitch-filter is actually usable by default).
3. Fixed the RST indentation and kept the concrete DTS examples in migration-guide-4.4.rst so they render properly and clearly state that Pulse mode 0 is unsupported and you must use prescale-glitch-filter-bypass;.
4. Updated the binding text to be explicit about Pulse mode 0 and the required bypass property in nxp,lptmr.yaml.

zejiang0jason commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -93,6 +93,139 @@ Controller Area Network (CAN)
+    .. code-block:: devicetree
+
+       lptmr0: counter@40040000 {
+           compatible = "nxp,lpc-lptmr";
```
`nxp,lptmr`

Holt-Sun replied:
Good catch. Thank you.
